benchmarks/synchrobench/src/main/java/com/yahoo/oak/OffHeapList.java

```diff
@@ -19,114 +23,124 @@
-                    return MyBuffer.compareBuffers((MyBuffer) o1, (MyBuffer) key2);
-                } else {
-                    return MyBuffer.compareBuffers((MyBuffer) o1, (ScopedReadBuffer) key2);
-                }
+            if (o1 instanceof OffHeapList.Cell) {
+                o1 = ((Cell) o1).key.get();
+            }
+
+            if (o2 instanceof OffHeapList.Cell) {
+                o2 = ((Cell) o2).key.get();
+            }
 
-            } else if (o2 instanceof MyBuffer) {
-                Cell cell1 = (Cell) o1;
-                Object key1 = cell1.key.get();
-                if (key1 instanceof MyBuffer) {
-                    return MyBuffer.compareBuffers((MyBuffer) key1, (MyBuffer) o2);
+            if (o1 instanceof OakScopedReadBuffer) {
+                if (o2 instanceof OakScopedReadBuffer) {
+                    return keyGen.compareSerializedKeys((OakScopedReadBuffer) o1, (OakScopedReadBuffer) o2);
                 } else {
-                    return -1 * MyBuffer.compareBuffers((MyBuffer) o2, (ScopedReadBuffer) key1);
+                    // Note the inversion of arguments, hence sign flip
+                    return (-1) * keyGen.compareKeyAndSerializedKey((BenchKey) o2, (OakScopedReadBuffer) o1);
                 }
-            } else if (o1 instanceof OffHeapList.Cell && o2 instanceof OffHeapList.Cell) {
-                Cell cell1 = (Cell) o1;
-                Object key1 = cell1.key.get();
-                Cell cell2 = (Cell) o2;
-                Object key2 = cell2.key.get();
-
-                if (key1 instanceof MyBuffer && key2 instanceof MyBuffer) {
-                    return MyBuffer.compareBuffers((MyBuffer) key1, (MyBuffer) key2);
-                } else if (key1 instanceof ScopedReadBuffer && key2 instanceof ScopedReadBuffer) {
-                    return MyBuffer.compareBuffers((ScopedReadBuffer) key1, (ScopedReadBuffer) key2);
-                } else if (key1 instanceof MyBuffer && key2 instanceof ScopedReadBuffer) {
-                    return MyBuffer.compareBuffers((MyBuffer) key1, (ScopedReadBuffer) key2);
+            } else {
+                if (o2 instanceof OakScopedReadBuffer) {
+                    return keyGen.compareKeyAndSerializedKey((BenchKey) o1, (OakScopedReadBuffer) o2);
                 } else {
-                    return -1 * MyBuffer.compareBuffers((MyBuffer) key2, (ScopedReadBuffer) key1);
+                    return keyGen.compareKeys((BenchKey) o1, (BenchKey) o2);
                 }
-            } else {
-                throw new UnsupportedOperationException();
             }
         };
+    }
 
+    @Override
+    public void init() {
         skipListMap = new ConcurrentSkipListMap<>(comparator);
-        allocator = new NativeMemoryAllocator(OAK_MAX_OFF_MEMORY);
-        mm = new SyncRecycleMemoryManager(allocator);
+        allocator = new NativeMemoryAllocator(MAX_OFF_MEMORY);
+        mm = new SeqExpandMemoryManager(allocator);
```

Comment:
`OffHeapList` didn't really use the MM. It just uses it to generate the slice. I changed it because `OffHeapList` doesn't use the header anyway, and `SeqExpandMemoryManager` doesn't have a header.

Reply:
Then how the compute works? `ScopedWriteBuffer.compute(prevValue.value.get().getSlice(), computeFunction);` Locks of MM are never used?

If so please add a comment about it

Reply:
compute/serialize or any of the writing functions does not use the locks since we only write in a scoped session (i.e., `ScopedWriteBuffer`). The lock is obtained beforehand by the map.
`OffHeapList` uses a synchronized block to avoid data races.
I'll add a comment mentioning why we use `SeqExpandMemoryManager` in this context.